### ISSUE TITLE
Problem indexing various indices

### ISSUE THREAD

**Issue by sonic182** [meilisearch/MeiliSearch#1196]
Hi

**Describe the bug**
I tried to index 55 index from customers where I work, the 55 most searched index from one of our ES clusters, they have various schemas, length of index, size of documents, etc... The problem is that meilisearch get stucks, it stop responding and also the web gui, seems like a bug somewhere.

**To Reproduce**
I used the next docker-compose for meilisearch

```yaml
version: "3.9"
   
services:
  es56:
    image: getmeili/meilisearch
    ports:
      - 7700:7700
    command: ['./meilisearch', '--http-addr', '0.0.0.0:7700', '--http-payload-size-limit', '1000000000', '--no-sentry', '1']
    # command: ['./meilisearch', '--env', 'production', '--http-addr', '0.0.0.0:7700', '--http-payload-size-limit', '1000000000', '--no-sentry', '1', '--master-key', '12341234']

networks:
  default:
    external:
      name: apps
```

Python script used to index files

```python
import logging
import json
import gzip
import os
from argparse import ArgumentParser
from contextlib import contextmanager

import meilisearch


@contextmanager
def open_file(filename, mode):
    if filename.endswith('.gz'):
        with gzip.GzipFile(filename, mode) as _file:
            yield _file
    else:
        with open(filename, mode) as _file:
            yield _file


def import_to_meilisearch(filesource: str, meili_server: str, indexname: str):
    client = meilisearch.Client(meili_server, '12341234')

    # An index is where the documents are stored.
    index = client.index(indexname)

    with open_file(filesource, 'rb') as _file:
        items_list = json.load(_file)

    count = 0
    chunk_size = 100
    chunk = []
    for item in items_list:
        count += 1
        chunk.append(item)

        if count % chunk_size == 0:
            logging.info(
                f'Processed {count} items of {indexname}, {len(chunk)}')

        # to make sure al prices are float, if needed.
        # for key in item.keys():
        #     if key.endswith('price'):
        #         try:
        #             item[key] = float(item[key])
        #             print(item[key])
        #         except Exception:
        #             logging.exception('error parsing')
        #             item[key] = 0.0
        #             print(item[key])

        if len(chunk) == chunk_size:
            index.add_documents(chunk)
            chunk = []

    if len(chunk) > 0:
        index.add_documents(chunk)
    logging.info(f'inserted {count} index')


def get_args():
    parser = ArgumentParser()
    parser.add_argument(
        'filesfolder', type=str, help='feed or path to import.')
    parser.add_argument(
        'meili-server', type=str, help='MeiliSearch server target')
    return parser.parse_args()


def start(args: ArgumentParser):
    if args.folder:
        filenames = os.listdir(args.filesfolder)

        for filename in filenames:
            _prefix, indexname = filename.split('_')
            # all files are in gzip...
            indexname = indexname.strip('.json.gz')
            filename = f'{args.file.strip("/")}/{filename}'

            import_to_meilisearch(filename, args.meili_server, indexname)


def main():
    logging.basicConfig(level=logging.INFO)

    args = get_args()
    start(args)


main()
```

Of course I can't give the files, but maybe getting various public datasets it can be reproduced. json file structures: `[{"foo": "bar", ...} {"bar": "baz", ...}]`

**Expected behavior**
To index all data ok

**Desktop (please complete the following information):**
 - OS: Ubuntu 18, Docker version 20.10.2, build 2291f61
 - docker meili latest image (0.18)

**Comment by davidhorat** [comment]
Same problem here.

**Comment by MarinPostma** [comment]
may be related to #1178, thanks for reporting!

**Comment by fxi** [comment]
Everything worked fine until I tried to quickly update stuff with a lot of indexes which is closest to what I would expect from a production use case. Is that parts of your tests ?

The workaround for me was to add a 10 second delay before each `addDocuments`. I also rewrote my logic: instead of updating the indexes as soon as a new document is available, I defined a background job that reindexes every hour. It's not very clean, but I couldn't find a better way for our prototype. We don't have many documents <3000, but parts of each document should be present in 10 different indexes, one per supported language, which trigger a lot of update in some cases.

As `.../updates` or `/health` are stuck too, I had to wrap a `Promise.race` + `setTimeout` to "guess" if meili isn't stuck from the client side. But if it does, then, what..

### Docker issue

This type of "blocking situation" should bring down the service, I think. If you are using meili with docker this should stop the container and thus docker could trigger a restart if `restart-policy` is `always`. You could add In your Dockerfile, something like this, maybe : 

```Dockerfile
HEALTHCHECK --interval=5m --timeout=3s --start-period=45s \
    CMD curl -f "http://localhost:7700/health" \
    || kill 1
# ps 1 being tini
```

### Roadmap ?

Is there something on your roadmap ? I should present a working prototype in .. two weeks and it's kind of work now, but.. it's not very clean.

I tried `v0.21.0-alpha.2` but it was not a huge success.


Thanks !

**Comment by MarinPostma** [comment]
Hey @fxi,

Thanks for the feedback! Can you explain a bit further what you are doing, and what are the result you are obtaining? Maybe a small script that would allow me to reproduce?

Generally speaking, I would suggest batch indexing once in a while.

**Comment by fxi** [comment]
I am trying to update the indexes in batches and Meili stops responding in some cases. Similar to what @sonic182 was describing. 

I'm not sure what's going on. No error log in the Meili service, it's just stuck.

I created a very rudimentary application to retrieve 5000 documents and create, then fill or update 10 Meili indexes.

Host and key can be configured.

https://jsfiddle.net/fxi/zc1jqtkd/

By default I put my own test machine and the private key is visible, but the localhost instance should work as well. This instance will be down in a few days.

So, conclusion ... After many (many) updates, Meili was still operational ... My problem was not reproducible locally (docker instance) or remotely (native on debian).

I am a little confused. I will continue to investigate. Maybe something in my real indexes is causing Meili to fail ...

**Comment by fxi** [comment]
### Synonyms 📚 ?

I think I found was was blocking Meili in my case.  Synonyms. I was not suspecting this.
```json
{
  "ABW": ["Aruba"],
  "Aruba": ["ABW"],
  "AFG": ["Afghanistan"],
  "Afghanistan": ["AFG"],
  "AGO": ["Angola"],
  "Angola": ["AGO"]
  [...]
}
```
For each index, one per language, I've added a synonym list for common location names : countries, and region. So, a search on ABW would return Aruba stuff. And Aruba search would return documents tagged "ABW".
This step was adding a list of 562 pairs. 
I removed it and Meili seems to fly again 🚀

### But ! Native version ok, docker version not ok 😱 :

In my NEW example script : 
https://jsfiddle.net/fxi/zc1jqtkd/5/
I've uploaded the paired synonyms list and Meili on the remote machine (test.fxi.io) did not flinch. 👍
But, the local dockerised, version (e.g. http://localhost:7700 ), the machine is stuck forever. 💩

So, what could it be ? Ressource ?  I don't think..

![image](https://user-images.githubusercontent.com/1196833/114915110-7346dd00-9e23-11eb-950c-21b0f19b9419.png)

**Comment by sonic182** [comment]
> ### Synonyms books ?
> 
> I think I found what was blocking Meili in my case. Synonyms. I was not suspecting this.
> 
> ```json
> {
>   "ABW": ["Aruba"],
>   "Aruba": ["ABW"],
>   "AFG": ["Afghanistan"],
>   "Afghanistan": ["AFG"],
>   "AGO": ["Angola"],
>   "Angola": ["AGO"]
>   [...]
> }
> ```
> 
> For each index, one per language, I've added a synonyms list for common location names : countries, and region. So, a search on ABW would return Aruba stuff. And Aruba search would return documents tagged "ABW".
> This step was adding a list of 562 pairs.
> I removed it and Meili seems to fly again rocket
> ### But ! Native version ok, docker version not ok scream :
> 
> In my NEW example script :
> https://jsfiddle.net/fxi/zc1jqtkd/5/
> I've uploaded the paired synonyms list and Meili on the remote machine (test.fxi.io) did not flinch. +1
> But, the local dockerised, version (e.g. http://localhost:7700 ), the machine is stuck forever. hankey
> 
> So, what could it be ? Ressource ? I don't think..
> 
> ![image](https://user-images.githubusercontent.com/1196833/114915110-7346dd00-9e23-11eb-950c-21b0f19b9419.png)

This kind of errors happens in c/c++ development, when using threads and not being good at synchronization hehe, This is very weird to happen in rust

**Comment by fxi** [comment]
Mhhé. 

Failure after failure... It's still blocking after a synonyms update...

I thought this could be an issue with alpine. So, I've built an image for `scratch`

```Dockerfile
ARG VERSION=0.20.0
ARG REPO=https://github.com/meilisearch/MeiliSearch/archive/refs/tags/
FROM rust:1.51.0 AS build
ARG VERSION
ARG REPO
RUN rustup target add x86_64-unknown-linux-musl
RUN apt update && apt install -y musl-tools
WORKDIR /build
WORKDIR /build/meilisearch
RUN curl -L ${REPO}/v${VERSION}.tar.gz | tar xzv
WORKDIR MeiliSearch-${VERSION}
RUN echo "1" && cargo build --release --target x86_64-unknown-linux-musl
RUN cargo build --release 
RUN mkdir -p /built
RUN cp target/x86_64-unknown-linux-musl/release/meilisearch /built/
FROM scratch
COPY --from=build /built/meilisearch .
ENV MEILI_HTTP_ADDR 0.0.0.0:7700
EXPOSE 7700/tcp
USER 1000
CMD ["./meilisearch"]
```
Exact same thing happens. Then, I thought, this could be better on `minideb`, so ...

```Dockerfile
ARG VERSION=0.20.0
ARG REPO=https://github.com/meilisearch/MeiliSearch/archive/refs/tags/
FROM rust:1.51.0 AS build
ARG VERSION
ARG REPO
WORKDIR /build
WORKDIR /build/meilisearch
RUN curl -L ${REPO}/v${VERSION}.tar.gz | tar xzv
WORKDIR MeiliSearch-${VERSION}
RUN cargo build --release 
RUN mkdir -p /built
RUN cp target/release/meilisearch /built/
FROM bitnami/minideb:jessie
COPY --from=build /built/meilisearch .
ENV MEILI_HTTP_ADDR 0.0.0.0:7700
EXPOSE 7700/tcp
USER 1000
CMD ["./meilisearch"]
```

Exact same thing. Any hint ?

**Comment by MarinPostma** [comment]
I need to investigate that

**Comment by fxi** [comment]
Should I start to learn how to debug a Rust app myself  ? haha 

This issue is also reproducible easily with the macOS version (v0.20.0) installed with `brew install meilisearch`

1) `brew install meilisearch`
2) launch `meilisearch` ->> http://127.0.0.1:7700
3) Load the test app : https://jsfiddle.net/fxi/zc1jqtkd/5
4) Configure the host as `http://127.0.0.1:7700`  ( key field can be empty if no key is set )
5) Click on update : after two indexes, it's stuck forever

Same thing happens with

- docker service inside a digital ocean droplet.  e.g. https://meilirocks.fxi.io (which will be stuck forever if synonyms  are updated ... ) NOTE: If you provide a RSA key, I can give you access to this test server.
- empty debian docker image with docker binary from your releases. So, non-compiled by me.

Thanks :)

Here is a picture my cat, which should bring you joy :
![image](https://user-images.githubusercontent.com/1196833/115589038-aed42200-a2cf-11eb-8ea0-019d5c15a0cb.png)

**Comment by curquiza** [comment]
Hello @fxi!

Thanks for helping us debug! 🙂

I tried your steps locally. It's weird because MeiliSearch is indeed stuck when calling routes like `GET /indexes/` or `GET /indexes/:uid/debug_en`, but I had no issue calling the `/health` or the `/version` route: MeiliSearch was able to respond.
I need to stop and re-launch meilisearch so that I can make it work again.

We will try to investigate it with the alpha versions, but the synonyms are not implemented at the moment.

Thanks again

PS: your cat is really lovely 😸

**Comment by volkv** [comment]
Can confirm such "stucks" when batch importing data to 5+ indexes (I'am using it with Laravel Scout).
Adding a delay of 5-10 seconds between indexes processing helps :)
And thanks for this awesome engine, I'am now migrating from Elastic, with Scout 9 native Meili support ;)

**Comment by montaniasystemab** [comment]
We're also experiencing this with both native and docker version of meilisearch, after creating four indexes with a couple of hundred to thousands of documents each everything just stops, the example code from the readme file will also be stuck forever:

```
$ curl -i -X POST 'http://127.0.0.1:7700/indexes' --data '{ "name": "Movies", "uid": "movies" }'
(hangs forever)
``` 

But will happily tell you that the service is fine: 

```
$ curl  127.0.0.1:7700/health
{"status":"available"}
```

I will try to add a delay between each document and index and see if that makes any difference.

Meilisearch from local docker reports this version: 
```
meili_1      | Build date:              "2021-03-30T15:13:32.344119665+00:00"
meili_1      | Package version: "0.20.0"
```

**Comment by curquiza** [comment]
Hello @montaniasystemab @volkv @fxi @sonic182 and everyone following the issue!

The RC of the v0.21.0 is out! https://github.com/meilisearch/MeiliSearch/releases

I tested it with your example @fxi (cf [your comment](https://github.com/meilisearch/MeiliSearch/issues/1196#issuecomment-824203846)) and I don't have the bug anymore!

Can you confirm it's the same for you?

**Comment by fxi** [comment]
Hello !

Thanks :)

It seems to work well using the docker image and the test script

```sh
docker run --rm -e MEILI_MASTER_KEY=thankscurquiza -p 7700:7700 getmeili/meilisearch:v0.21.0rc1 ./meilisearch
```
![meilisearch_0 21 0rc1](https://user-images.githubusercontent.com/1196833/124240234-d7bc3300-db1a-11eb-91b3-fe6cc6a1b577.gif)

**Comment by sonic182** [comment]
Gonna try it at Monday or next sprint in my job :)

Thanks!

**Comment by montaniasystemab** [comment]
Using v0.21.0rc1 we're no longer able to reproduce the issue from v0.20 :+1: 

Thanks!

**Comment by curquiza** [comment]
Thanks a lot for your returns! 🙏 I close this issue then!! It will be released in the v0.21.0 in a few weeks
